common/src/graph.rs

```diff
@@ -159,62 +159,39 @@ impl Graph {
     /// Ensures that the neighbour node at a particular side of a particular node exists in the graph,
     /// as well as the nodes from the origin to the neighbour node.
     pub fn ensure_neighbor(&mut self, node: NodeId, side: Side) -> NodeId {
-        let v = &self.nodes[&node];
-        if let Some(x) = v.neighbors[side as usize] {
-            // Neighbor already exists
-            return x;
-        }
-
-        let neighbor_is_further = |parent_side| {
-            (side != parent_side && !side.adjacent_to(parent_side))
-                || !self.is_near_side(v.parent().unwrap(), side)
-        };
-
-        // Create a new neighbor
-        if v.parent_side.map_or(true, neighbor_is_further) {
-            // New neighbor will be further from the origin
-            return self.insert_child(node, side);
-        }
-
-        // Neighbor is closer to the origin; find it, backfilling if necessary
-        let x = self.nodes[&v.parent().unwrap()].neighbors[side as usize].unwrap();
-        let parent_side = v.parent_side.unwrap();
-        let neighbor = self.ensure_neighbor(x, parent_side);
-        self.link_neighbors(node, neighbor, side);
-        neighbor
+        self.nodes[&node].neighbors[side as usize]
+            .unwrap_or_else(|| self.insert_neighbor(node, side))
     }
 
     /// Whether `node`'s neighbor along `side` is closer than it to the origin
-    fn is_near_side(&self, node: NodeId, side: Side) -> bool {
+    fn is_descender(&self, node: NodeId, side: Side) -> bool {
         let v = &self.nodes[&node];
         v.neighbors[side as usize].map_or(false, |x| self.nodes[&x].length < v.length)
     }
 
-    pub fn insert_child(&mut self, parent: NodeId, side: Side) -> NodeId {
-        // Always create shorter nodes first so that self.nodes always puts parent nodes before their child nodes, enabling
-        // graceful synchronization of the graph
-        let shorter_neighbors = self.populate_shorter_neighbors_of_child(parent, side);
-
-        // Select the side along the canonical path from the origin to this node. Future work: make
-        // this the parent to reduce the number of concepts.
-        let (path_side, predecessor) = shorter_neighbors
-            .clone()
-            .chain(Some((side, parent)))
-            .min_by_key(|&(side, _)| side)
-            .unwrap();
+    /// Inserts the neighbor of the given node at the given side into the graph, ensuring that all
+    /// shorter nodes it depends on are created first.
+    pub fn insert_neighbor(&mut self, node: NodeId, side: Side) -> NodeId {
+        // An assumption made by this function is that `side` is not a descender of `node`. This is a safe assumption to make
+        // because a node cannot be constructed before its shorter neighbors. The call to `populate_shorter_neighbors_of_neighbor`
+        // guarantees this.
+        let shorter_neighbors_of_neighbor = self.populate_shorter_neighbors_of_neighbor(node, side);
```

Comment:
I found this variable name confusing. I understood it to mean "shorter neighbors of the neighbor [of the new node]", but in fact it seems to be "shorter neighbors [of the new node]". Maybe standardize on `subject` like below?

Reply:
I'm not surprised it's confusing. It's been a while, so looking at this PR again with fresh eyes should hopefully help me think of better variable names here.

Reply:
Thinking on this again, I agree with you that standardizing on the name `subject` makes sense, as long as that naming decision doesn't leak outside the `graph` module. I'll go ahead with your suggestion.